src/mono/mono/mini/interp/transform.c

```diff
@@ -10188,61 +10171,96 @@ add_active_call (TransformData *td, ActiveCalls *ac, InterpInst *call)
+ *
+ * Call is push onto the stack with an array of other active calls that need to be resolved first.
+ *
+ * If there are no other active calls, function retrieves calls from the stack and resolves them, including the current call and all deferred calls.
+ *
+ * For each call, function computes the call offset of each call argument starting from a base offset, and stores the computed call offset into a InterpInst.
+ * The base offset is computed as max offset of all call offsets on which the call depends.
+ * Stack ensures that all call offsets on which the call depends are calculated before the call resolution.
+ */
 static void
 end_active_call (TransformData *td, ActiveCalls *ac, InterpInst *call)
 {
 	// Remove call from array
 	for (int i = 0; i < ac->active_calls_count; i++) {
-		if (ac->active_calls [i].call == call) {
+		if (ac->active_calls [i] == call) {
 			ac->active_calls_count--;
-			ac->param_size -= ac->active_calls [i].param_size;
 			// Since this entry is removed, move the last entry into it
 			if (ac->active_calls_count > 0 && i < ac->active_calls_count)
 				ac->active_calls [i] = ac->active_calls [ac->active_calls_count];
+			break;
 		}
 	}
-	// This is the relative offset (to the start of the call args stack) where the args
-	// for this call reside.
-	int start_offset = ac->param_size;
 
-	// Compute to offset of each call argument
-	int *call_args = call->info.call_args;
-	if (call_args && (*call_args != -1)) {
-		int var = *call_args;
-		while (var != -1) {
-			alloc_var_offset (td, var, &start_offset);
-			call_args++;
-			var = *call_args;
-		}
+	// Push active call that should be resolved onto the stack
+	call->info.call_info->call_deps = NULL;
+	if (ac->active_calls_count) {
+		for (int i = 0; i < ac->active_calls_count; i++)
+			call->info.call_info->call_deps = g_slist_prepend_mempool (td->mempool, call->info.call_info->call_deps, ac->active_calls [i]);
+		ac->deferred_calls = g_slist_prepend_mempool (td->mempool, ac->deferred_calls, call);
 	} else {
-		// This call has no argument. Allocate a dummy one so when we resolve the
-		// offset for MINT_CALL_ARGS_SREG during compacted instruction emit, we can
-		// always use the offset of the first var in the call_args array
-		int new_var = create_interp_local (td, mono_get_int_type ());
-		td->locals [new_var].call = call;
-		td->locals [new_var].flags |= INTERP_LOCAL_FLAG_CALL_ARGS;
-		alloc_var_offset (td, new_var, &start_offset);
-
-		call_args = (int*)mono_mempool_alloc (td->mempool, 3 * sizeof (int));
-		call_args [0] = new_var;
-		call_args [1] = -1;
+		// If no other active calls, current active call and all deferred calls can be resolved from the stack
+		InterpInst *deferred_call = call;
+		while (deferred_call) {
+			// `base_offset` is a relative offset (to the start of the call args stack) where the args for this call reside.
+			int base_offset = 0;
```

Comment:
I think a comment here explaining why `call_offset` is guaranteed to be set for all deps wouldn't hurt. Something like: `The deps for a call represent the list of active calls at the moment when the call ends. This means that all deps for a call end after the call in question. Given we iterate over the list of deferred calls from the last to the first one to end, all deps of a call are guaranteed to have been processed at this point.`

Reply:
Added.